### DISCUSSION TITLE
Support `CC_FOR_BUILD` in autoconf et al

### DISCUSSION THREAD

**Discussion by ras0219-msft** [microsoft/vcpkg#20755] [Ideas]
First, let me lay out the concrete example of `libgpg-error`.

I've also looked at moving `libgpg-error` to build without the SMP changes -- or even with the SMP changes but using the original buildsystem (similar to https://github.com/microsoft/vcpkg/pull/20447). By changing three more lines, I was able to get it to build using autoconf while native-targeting.

Unfortunately, `libgpg-error` is operating outside the bounds of what vcpkg can properly support at this time. It wants to:
1. Build and run an executable for the target to gather information
2. Use the information in (1) to build and run a code generator for the host
3. Use the output of (2) to build the final libraries for the target

See https://github.com/gpg/libgpg-error/blob/master/README#L119 for more details.

vcpkg does not currently support flowing information backwards from the target to the host -- it only supports flowing forward (via host dependencies). There are cached results from step (1) available in the repository for well known platforms, however the problem remains that step (2) implies building a separate host binary for every possible target. SMP caches the results of (2) for Windows and UWP which works around the issue for those specific platforms.

One way we could possibly work around this is by introducing a host port helper that captures the build environment for the host and saves it as a shell script. Then, consumers could invoke this "cc-host.bat"/"cc-host.sh" and it would restore the full environment (including PATH, INCLUDE, etc) as well as adding on any needed flags.

This would work perfectly except for binary caching. The contents of that file will be _extremely_ system sensitive and unrelocatable since it records absolute paths. We may be able to work around this using the undocumented features `VCPKG_PUBLIC_ABI_OVERRIDE`[1] and `vcpkg-abi-settings`[2] to produce a package that "depends upon" the precise value of those environment variables, but this is suspect at best. To fully support this, I believe the vcpkg tool will need dedicated feature work to support capturing / flattening the toolchain this way.

+@wrobelda +@Neumann-A +@dg0yt for input since I think you've all touched autoconf recently

[1]: https://github.com/microsoft/vcpkg-tool/blob/909c2ba3d9e2de8fd3381d7326885877f795ecc1/src/vcpkg/build.cpp#L1471
[2]: https://github.com/microsoft/vcpkg-tool/blob/909c2ba3d9e2de8fd3381d7326885877f795ecc1/src/vcpkg/cmakevars.cpp#L348

**Comment by wrobelda** [comment]
@ras0219-msft I looked into it today and I think we don't need to go this deep.

There is currently a problem with cross-compilation, as you are aware. However, per their instructions you quoted, my understanding is that it should be possible to avoid  steps 1 and 2 you described, for as long as they include that information with the code upstream. It is not, however, currently possible to do that for Windows: https://dev.gnupg.org/T5687#151837

This is essentially something they should hopefully be able to fix upstream. Unfortunately, due to the nature of their non-standard set-up, they stand out as one of the few modern-day libraries having so many cross-compilation issues, and "it is what it is".[^1][^2]

Another issue is that consumers of `libgpg-error` tend to depend on the `gpgerror-config` script, but apparently this is optional now, since [they added support for pkg-config](https://dev.gnupg.org/T4085): https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=643341#100 (this whole thread is a good read, BTW). 

**EDIT**: Well, indeed, they [deprecate](https://dev.gnupg.org/T5683) the `*-config` scripts, and using pkg-config only should work, except it doesn't with vcpkg (tested it with `libgcrypt`), and that's because they came up with [another set of custom magic](https://dev.gnupg.org/T5365#144688) for locating the `.pc` files. I [explained](https://dev.gnupg.org/T5365#151842) why is this problematic, hopefully it gets fixed soon.
      
[^1]: https://bugs.gentoo.org/584052#c50
[^2]: https://dev.gnupg.org/T2370#131742

**Comment by MouriNaruto** [comment]
@ras0219-msft 

I think add the real support for `CC_FOR_BUILD` in autoconf with MSVC can solve cross compiling issues. (For example, I want to compile arm version of fribidi, but it failed because the `CC_FOR_BUILD` points to the msvc arm compiler, it use `CC_FOR_BUILD` to build some generators and the Windows x64 don't support arm binaries so it can't compile properly.)

Kenji Mouri

**Comment by Neumann-A** [comment]
I think I hit something similar. If we have a host dependency on `pkgconf` it needs to be setup to actually be a host dependency targeting the target, so that it can find target system dependencies installed by the user. So the port needs a mix of host and target information to be correctly setup. Maybe we can mark such ports in the manifest with something like `host_for_target: true` which marks such ports as special and they get additional information and an extra installation directory being `${HOST_TRIPLET}_${TARGET_TRIPLET}`. Another possibility is that we require that the `--host-triplet`  is in reality a specialized triplet for the target but setup to build on the host. This however might require clarifying that in the docs.  

I don't see VCPKG support `CC_FOR_BUILD`. In general such MAKEFILES need to be patched and the executable externally provided. In the case of `libgpg-error` there is nothing we can do except splitting the port for the 1. step and the rest and provide a variable to insert the result of  `ssh TARGET ./gen-posix-lock-obj >tmp.h` / or accept different result.h headers in the porfile.  The reason for that is that vcpkg has no concept on how to run stuff for the target. 

If vcpkg learns some way to run stuff for the target we can assume that `--host-triplet`  will be a very specialized triplet for the target but setup to build on the host and as such will contain information about the target and will know how to run stuff for it. This then circumvents the need to enable `CC_FOR_BUILD` since the host triplet will be able to get the required information and forward it to the target triplet without the need for the target triplet to build for the host.  

TL;DR: `--host-triplet` will probably not be a generalized triplet for the host but a triplet for the host considering the target.
